### PR TITLE
Add documentation, examples and tests

### PR DESCRIPTION
**Pull request by hug-dev**
Signed-off-by: Hugues de Valon <hugues.devalon@arm.com>

### CODE REVIEW

File: psa-crypto/src/operations/asym_signature.rs [hunk 1]
```diff
@@ -2,13 +2,57 @@
...
+/// #     policy: Policy {
+/// #         usage_flags: UsageFlags {
+/// #             sign_hash: true,
+/// #             sign_message: true,
+/// #             verify_hash: true,
+/// #             verify_message: true,
+/// #             ..Default::default()
+/// #         },
+/// #         permitted_algorithms: AsymmetricSignature::RsaPkcs1v15Sign {
+/// #             hash_alg: Hash::Sha256.into(),
```
Comment (egrimley-arm) [inline psa-crypto/src/operations/asym_signature.rs:37]: Why do this line and the following line contain `#`?

File: psa-crypto/src/operations/asym_signature.rs [hunk 2]
```diff
@@ -38,7 +82,49 @@ pub fn sign_hash(
     Ok(signature_length)
 }
 
-/// Verify a hash
+/// Verify the signature of a hash or short message using a public key
+///
+/// # Example
+///
+/// ```
+/// # use psa_crypto::operations::key_management::generate;
```
Comment (egrimley-arm) [inline psa-crypto/src/operations/asym_signature.rs:90]: Ditto.

File: psa-crypto/src/types/key.rs [hunk 3]
```diff
@@ -35,6 +35,39 @@ impl Attributes {
     }
 
     /// Check export in a faillible way
```
Comment (egrimley-arm) [inline psa-crypto/src/types/key.rs:37]: Not part of this patch, but while you're there: faillible -> fallible

File: psa-crypto/src/operations/asym_signature.rs (hunk 1 above)
Comment (hug-dev) [reply psa-crypto/src/operations/asym_signature.rs:37]: It is made to hide portions of the example so it only shows relevent parts. See [here](https://doc.rust-lang.org/rustdoc/documentation-tests.html#hiding-portions-of-the-example) for details.

File: psa-crypto/src/types/algorithm.rs [hunk 4]
```diff
@@ -42,6 +42,17 @@ pub enum Algorithm {
...
 impl Algorithm {
     /// Check if the algorithm is a HMAC algorithm, truncated or not
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use psa_crypto::types::algorithm::{Algorithm, Mac, FullLengthMac, Hash};
+    /// let hmac = Algorithm::Mac(Mac::Truncated {
+    ///     mac_alg: FullLengthMac::Hmac { hash_alg: Hash::Sha256 },
+    ///     mac_length: 34,
```
Comment (ionut-arm) [inline psa-crypto/src/types/algorithm.rs:52]: I'm assuming the `mac_length` is in bytes, not bits, in which case this is invalid because `Sha256` has only 32 bytes (not that I'm saying we shouldn't allow that construction, just saying that maybe something below 32 would be better for an example)

File: psa-crypto/src/types/algorithm.rs [hunk 5]
```diff
@@ -685,3 +720,42 @@ impl From<AsymmetricSignature> for psa_crypto_sys::psa_algorithm_t {
...
+                .unwrap()
+        );
+        assert_eq!(
+            psa_crypto_sys::PSA_ALG_ECDSA(psa_crypto_sys::PSA_ALG_SHA3_512),
+            Algorithm::AsymmetricSignature(AsymmetricSignature::Ecdsa {
+                hash_alg: SignHash::Specific(Hash::Sha3_512),
+            })
+            .try_into()
+            .unwrap()
+        );
```
Comment (ionut-arm) [inline psa-crypto/src/types/algorithm.rs:759]: Could you add some checks that fail (i.e. with `unwrap_err()`)?

File: psa-crypto/src/types/key.rs [hunk 6]
```diff
@@ -390,6 +469,23 @@ pub struct UsageFlags {
...
+            encrypt: false,
+            decrypt: false,
+            sign_message: false,
+            verify_message: false,
+            sign_hash: false,
+            verify_hash: false,
+            derive: false,
+        }
+    }
+}
```
Comment (ionut-arm) [inline psa-crypto/src/types/key.rs:487]: Is this not what deriving `Default` would obtain?

File: psa-crypto/src/types/key.rs [hunk 7]
```diff
@@ -1017,4 +1114,63 @@ mod tests {
...
+                    },
+                    permitted_algorithms: AsymmetricSignature::Ecdsa {
+                        hash_alg: Hash::Sha256.into(),
+                    }
+                    .into(),
+                },
+            },
+            attrs.try_into().unwrap()
+        );
+    }
```
Comment (ionut-arm) [inline psa-crypto/src/types/key.rs:1175]: 👌 nice

File: psa-crypto/src/operations/key_management.rs [hunk 8]
```diff
@@ -8,8 +8,44 @@ use crate::types::key::{Attributes, Id};
 use crate::types::status::{Result, Status};
 use core::convert::TryFrom;
 
-/// Generate a key
-pub fn generate_key(attributes: Attributes, id: Option<u32>) -> Result<Id> {
```
[the remark this answers is not in the document]
Comment (hug-dev) [reply psa-crypto/src/operations/key_management.rs:12]: > How come you dropped the _key part of the name, because it's in key_management? 

Yes, that's the reason!

> Then you probably need to modify the examples to call it as key_management::generate(...), because it's confusing as to what generate is and where it is coming from.

Sure, will do.

File: psa-crypto/src/types/algorithm.rs (hunk 4 above)
Comment (hug-dev) [reply psa-crypto/src/types/algorithm.rs:52]: Makes sense, will modify!

File: psa-crypto/src/types/algorithm.rs (hunk 5 above)
Comment (hug-dev) [reply psa-crypto/src/types/algorithm.rs:759]: Yes, will add!

File: psa-crypto/src/types/key.rs (hunk 6 above)
Comment (hug-dev) [reply psa-crypto/src/types/key.rs:487]: Oh yes, you are right, I forgot that you can derive `Default`

File: psa-crypto/src/operations/key_management.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (ionut-arm) [reply psa-crypto/src/operations/key_management.rs:12]: You've only modified the examples in this file, the other still import the function directly 🤡 but nevermind, we can discuss under #17
